azure-devops/create-new-agent-image.ps1

```diff
@@ -0,0 +1,297 @@
+  -Activity $ProgressActivity `
+  -Status 'Waiting for VM to shut down' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+Start-WaitForShutdown -ResourceGroupName $ResourceGroupName -Name $ProtoVMName
+
+####################################################################################################
+Write-Progress `
+  -Activity $ProgressActivity `
+  -Status 'Converting VM to Image' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+Stop-AzVM `
+  -ResourceGroupName $ResourceGroupName `
+  -Name $ProtoVMName `
+  -Force
+
+Set-AzVM `
+  -ResourceGroupName $ResourceGroupName `
+  -Name $ProtoVMName `
+  -Generalized
+
+$ImageConfig = New-AzImageConfig -Location $Location -SourceVirtualMachineId $VM.ID
+$Image = New-AzImage -Image $ImageConfig -ImageName $ProtoVMName -ResourceGroupName $ResourceGroupName
+
+####################################################################################################
+Write-Progress `
+  -Activity $ProgressActivity `
+  -Status 'Deleting unused VM and disk' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+Remove-AzVM -Id $VM.ID -Force
+Remove-AzDisk -ResourceGroupName $ResourceGroupName -DiskName $PrototypeOSDiskName -Force
+
+####################################################################################################
+Write-Progress `
+  -Activity $ProgressActivity `
+  -Status 'Creating scale set' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+$VmssIpConfigName = $ResourceGroupName + 'VmssIpConfig'
+$VmssIpConfig = New-AzVmssIpConfig -SubnetId $Nic.IpConfigurations[0].Subnet.Id -Primary -Name $VmssIpConfigName
+$VmssName = $ResourceGroupName + 'Vmss'
+$Vmss = New-AzVmssConfig `
+  -Location $Location `
+  -SkuCapacity 2 `
+  -SkuName $VMSize `
+  -SkuTier 'Standard' `
+  -Overprovision $false `
+  -UpgradePolicyMode Manual `
+  -EvictionPolicy Delete `
+  -Priority Spot `
+  -MaxPrice -1
+
+$Vmss = Add-AzVmssNetworkInterfaceConfiguration `
+  -VirtualMachineScaleSet $Vmss `
+  -Primary $true `
+  -IpConfiguration $VmssIpConfig `
+  -NetworkSecurityGroupId $NetworkSecurityGroup.Id `
+  -Name $NicName
+
+$Vmss = Set-AzVmssOsProfile `
+  -VirtualMachineScaleSet $Vmss `
+  -ComputerNamePrefix $LiveVMPrefix `
+  -AdminUsername 'AdminUser' `
+  -AdminPassword $AdminPW `
+  -WindowsConfigurationProvisionVMAgent $true `
+  -WindowsConfigurationEnableAutomaticUpdate $true
+
+$Vmss = Set-AzVmssStorageProfile `
+  -VirtualMachineScaleSet $Vmss `
+  -OsDiskCreateOption 'FromImage' `
+  -OsDiskCaching ReadWrite `
+  -ImageReferenceId $Image.Id `
+  -ManagedDisk Premium_LRS
+
+New-AzVmss `
+  -ResourceGroupName $ResourceGroupName `
+  -Name $VmssName `
+  -VirtualMachineScaleSet $Vmss
+
+####################################################################################################
+Write-Progress -Activity $ProgressActivity -Status -Completed
+Write-Reminders $AdminPW
+Write-Output 'Finished! Terminate.'

```

Comment:
Terminate sounds a little ominous, do we need that? END OF LINE.

Reply:
I always like little helper tools like this to emit that message but no skin off my nose to remove it. Done.

Reply:
“Finished” is fine 😸